pandas/tests/io/test_sql.py

```diff
@@ -1911,1102 +1818,1370 @@ def test_api_read_sql_duplicate_columns(conn, request):
+    # full NaN column (numeric float column)
+    conn = request.getfixturevalue(conn)
+    df = DataFrame({"A": [0, 1, 2], "B": [np.nan, np.nan, np.nan]})
+    assert df.to_sql(name="test_nan", con=conn, index=False) == 3
+
+    # with read_table
+    result = sql.read_sql_table("test_nan", conn)
+    tm.assert_frame_equal(result, df)
+
+    # with read_sql -> not type info from table -> stays None
+    df["B"] = df["B"].astype("object")
+    df["B"] = None
+    result = sql.read_sql_query("SELECT * FROM test_nan", conn)
+    tm.assert_frame_equal(result, df)
+
+
+@pytest.mark.parametrize("conn", sqlalchemy_connectable)
+def test_nan_string(conn, request):
+    # NaNs in string column
+    conn = request.getfixturevalue(conn)
+    df = DataFrame({"A": [0, 1, 2], "B": ["a", "b", np.nan]})
+    assert df.to_sql(name="test_nan", con=conn, index=False) == 3
+
+    # NaNs are coming back as None
+    df.loc[2, "B"] = None
+
+    # with read_table
+    result = sql.read_sql_table("test_nan", conn)
+    tm.assert_frame_equal(result, df)
+
+    # with read_sql
+    result = sql.read_sql_query("SELECT * FROM test_nan", conn)
+    tm.assert_frame_equal(result, df)
+
+
+@pytest.mark.parametrize("conn", all_connectable)
+def test_to_sql_save_index(conn, request):
+    if "engine" in conn:
+        pytest.skip(reason="fails and hangs forever with engine")
+
+    conn_name = conn
+    conn = request.getfixturevalue(conn)
+    df = DataFrame.from_records(
+        [(1, 2.1, "line1"), (2, 1.5, "line2")], columns=["A", "B", "C"], index=["A"]
+    )
+
+    pandasSQL = pandasSQL_builder(conn)
+    tbl_name = "test_to_sql_saves_index"
+    assert pandasSQL.to_sql(df, tbl_name) == 2
+
+    if conn_name in {"sqlite_buildin", "sqlite_str"}:
+        ixs = sql.read_sql_query(
+            "SELECT * FROM sqlite_master WHERE type = 'index' "
+            f"AND tbl_name = '{tbl_name}'",
+            conn,
+        )
+        ix_cols = []
+        for ix_name in ixs.name:
+            ix_info = sql.read_sql_query(f"PRAGMA index_info({ix_name})", conn)
+            ix_cols.append(ix_info.name.tolist())
+    else:
         from sqlalchemy import inspect
 
-        insp = inspect(self.conn)
+        insp = inspect(conn)
 
         ixs = insp.get_indexes(tbl_name)
-        ixs = [i["column_names"] for i in ixs]
-        return ixs
+        ix_cols = [i["column_names"] for i in ixs]
 
-    def test_to_sql_save_index(self):
-        self._to_sql_save_index()
+    assert ix_cols == [["A"]]
 
-    def test_transactions(self):
-        self._transaction_test()
 
-    def test_get_schema_create_table(self, test_frame3):
-        # Use a dataframe without a bool column, since MySQL converts bool to
-        # TINYINT (which read_sql_table returns as an int and causes a dtype
-        # mismatch)
+@pytest.mark.parametrize("conn", all_connectable)
+def test_transactions(conn, request):
+    if "engine" in conn:
+        pytest.skip(reason="hangs forever in CI with engine")
```

Comment:
Is this still the case after the refactor?

Reply:
Same cases below.

Also, could you use `is_ci_environment()` in this condition too?